### PR TITLE
RFC: Add #[repr(pack = "N")]

### PR DESCRIPTION
**Pull request by retep998**
[Rendered](https://github.com/retep998/rfcs/blob/optimal-rabbit-packing/text/0000-repr-pack.md)

### CODE REVIEW

File: text/0000-repr-pack.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+parsed as a `u64`. The restrictions on where this attribute can be placed along
+with the accepted values are:
+
+* Custom packing can only be specified on `struct` declarations for now.
+  Specifying a different packing on perhaps `enum` or `type` definitions should
+  be a backwards-compatible extension.
+* Packing values must be a power of two.
+
+By specifying this attribute, the alignment of the struct would be the smaller
+of the specified packing and the default alignment of the struct otherwise. The
```
Comment (cmr) [inline text/0000-repr-pack.md]: s/otherwise//

File: text/0000-repr-pack.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+`#[repr(packed)]` and `#[repr(pack = "1")]` should have identical behavior.
+
+Because this lowers the effective alignment of fields in the same way that
+`#[repr(packed)]` does (which caused https://github.com/rust-lang/rust/issues/27060 ),
+while accessing a field should be safe, borrowing a field should be unsafe.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This would unfortunately make my life easier even though one of the unstated
```
Comment (cmr) [inline text/0000-repr-pack.md]: Please remove this paragraph.

File: text/0000-repr-pack.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  `#[repr(packed)]` with manual padding, although such structs would always have
+  an alignment of 1 which is often wrong.
+* Alternatively a new attribute could be used such as `#[pack]`.
+* `#[repr(packed)]` could be extended as either `#[repr(packed(N))]` or
+  `#[repr(packed = "N")]`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* The behavior specified here should match the behavior of MSVC at least. Does
```
Comment (cmr) [inline text/0000-repr-pack.md]: This must be resolved. Verify against the behavior of clang (or gcc).

File: text/0000-repr-pack.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  it match the behavior of other C/C++ compilers as well?
+* Should it still be safe to borrow fields whose alignment is less than or equal
```
Comment (cmr) [inline text/0000-repr-pack.md]: It should have the same behavior of `#[repr(packed)]` I think.

**Comment by cmr** [comment]
The feature is necessary. If this behavior is in line with common C compilers, :shipit:

File: text/0000-repr-pack.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+[summary]: #summary
+
+Extend the existing `#[repr]` attribute on structs with a `pack = "N"` option to
+specify a custom packing for `struct` types.
+
+# Motivation
+[motivation]: #motivation
+
+Many C/C++ compilers allow a packing to be specified for structs which
+effectivally lowers the alignment for a struct and its fields (for example with
```
Comment (cmr) [inline text/0000-repr-pack.md]: s/effectivally/effectively/

File: text/0000-repr-pack.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+MSVC there is `#pragma pack(N)`). Such packing is used extensively in certain
+C/C++ libraries (such as Windows API which uses it all over the place making
+writing Rust libraries such as `winapi` a nightmare).
```
Comment (cmr) [inline text/0000-repr-pack.md]: s/a nightmare/challenging/

File: text/0000-repr-pack.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+Extend the existing `#[repr]` attribute on structs with a `pack = "N"` option to
+specify a custom packing for `struct` types.
+
+# Motivation
+[motivation]: #motivation
+
+Many C/C++ compilers allow a packing to be specified for structs which
+effectivally lowers the alignment for a struct and its fields (for example with
+MSVC there is `#pragma pack(N)`). Such packing is used extensively in certain
+C/C++ libraries (such as Windows API which uses it all over the place making
```
Comment (cmr) [inline text/0000-repr-pack.md]: s/all over the place/pervasively/

File: text/0000-repr-pack.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+writing Rust libraries such as `winapi` a nightmare).
+
+At the moment the only way to work around the lack of a proper
+`#[repr(pack = "N")]` attribute is to use `#[repr(packed)]` and then manually
+fill in padding which is a burdensome task. Even then that isn't quite right
+because the overall alignment of the struct would end up as 1 even though it
+needs to be N (or the default if that is smaller than N), so this fills in a gap
+which is basically impossible to do in Rust at the moment.
```
Comment (cmr) [inline text/0000-repr-pack.md]: s/basically //

File: text/0000-repr-pack.md (hunk 3 above)
Comment (retep998) [inline text/0000-repr-pack.md]: Specifically clang or gcc on a platform other than Windows. On Windows they are supposed to match the msvc ABI.

**Comment by DanielKeep** [comment]
For reference, here's [an example of Clang doing this](http://goo.gl/PeYrl5).

**Comment by retep998** [comment]
See [his comment](https://github.com/rust-lang/rfcs/pull/1358#issuecomment-162403439) for an interesting interaction edge case with the alignment RFC regarding required alignment.

**Comment by nikomatsakis** [comment]
**Hear ye, hear ye!** This RFC is now entering **final comment period**.

**Comment by ubsan** [comment]
This seems right to have... although, I think it should be `#[repr(packed = "N")]`. So the default of `#[repr(packed)]` would be `#[repr(packed = "1")]`.

**Comment by tsion** [comment]
@ubsan `#[repr(packed(1))]` is not currently valid attribute syntax (but `#[repr(packed = "1")]` is). I guess the RFC mention this, so cc @retep998 as well.

It could become valid with https://github.com/rust-lang/rfcs/pull/1559, though, or as @nrc mentions in a comment, the procedural macros RFC.

**Comment by retep998** [comment]
I don't mind the syntax either way, as long as I can have this attribute at all.

**Comment by starkat99** [comment]
I think it should just extend `#[repr(packed)]` rather than adding a new `pack` syntax. The duplication is confusing otherwise. Just have `#[repr(packed)]` be default to `#[repr(packed = "1")]`.

**Comment by nikomatsakis** [comment]
Hmm, I think I too prefer `#[repr(packed = "N")]`. Otherwise, the RFC seems unobjectionable to me. It'd also be nice to approve https://github.com/rust-lang/rfcs/pull/1358, which covers very similar subject matter.

**Comment by nikomatsakis** [comment]
One area the RFC did not (explicitly) address -- or I missed it -- is what to do if there are multiple `packed` specifiers. It did say that `#[repr(pack = "2")]` and `#[repr(packed)]` cannot be combined, but it did not say anything about `#[repr(pack = "2")]` and `#[repr(pack = "4")]`. Presumably this is an error as well? (Unlike the proposed align RFC?)

Is there any reason one might want multiple specifiers? Like, independent macros might add tighter and tighter requirements? (However, unlike with alignment, it's not obvious to me whether you would want the max or min packing that was specified.)

**Comment by retep998** [comment]
I think if we did allow specifying the packing multiple times, that the right thing to do would be to use the smallest packing specified. The alignment of any given field is `min(default, packing)` so extending packing to be `min(packing1, packing2)` seems fairly natural.

As for the align RFC, there'd need to be a decision on how it interacts with packing. Whether it does the msvc style `__declspec(align(N))` with required alignment that cannot be lowered by packing, or the gcc style alignment attribute that can be lowered by packing.

**Comment by nikomatsakis** [comment]
**Huzzah!** The @rust-lang/lang team has decided **to accept** this RFC, with the caveat that the attribute be renamed to `packed` (and not `pack`). I will make this change when merging (I assume @retep998 you do not object.)

We discussed a few other things:

- We should (obviously) clarify the relationship to an align attribute. We're going to bring the align RFC into FCP, discussion can happen there. But in the @rust-lang/lang meeting, I think the general feeling was that if both attributes are present, then the specified align would overrule the alignment derived from packing (in other words, we would match the MSVC style). Another option is to forbid both options from being used simultaneously, but that seems unnecessary to me.

**Comment by nikomatsakis** [comment]
Tracking issue: https://github.com/rust-lang/rust/issues/33158

If you'd like to keep following the development of this feature, please subscribe to that issue, thanks! :)

**Pull request merged**